### PR TITLE
Generic Webhooks

### PR DESCRIPTION
**Pull request by lp4775**
##### SUMMARY
Implement Generic Webhooks

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API + UI

##### AWX VERSION
dlevel

##### ADDITIONAL INFORMATION
This PR allows jobs to be started from generic webhooks. This feature will make integrating AWX into existing workflows much easier. I have set the defaults to match that of BitBucket but the reference keys and endpoint url can both be overridden using json config in the payload:
```
"config" : {
    "reference_keys" : {
        "Merge Request Hook" : "commit_hash",
        "Branch Deleted" : "branch_name"
    },
    "api_endpoint" : "http://www.github.com/commits/1234/build/status"
}
```

It has been tested via Postman but the following curl command should test the functionality:
```
curl --location --request POST 'https://local.awx:8043/api/v2/job_templates/7/generic/' \
--header 'x-generic-token: hmac_with_secret' \
--header 'x-EVENT-KEY: repo:comment:added' \
--header 'Content-Type: application/json' \
--data-raw '{
  "commit": "da1560886d4f094c3e6c9ef40349f7d38b5d27d7",
  "config" : {
    "api_endpoint" : "http://www.google.com"
  }
}'
```

### CODE REVIEW

**Comment by lp4775** [comment]
@wenottingham removed lingering debugging lines

**Comment by shanemcd** [comment]
Hello - thanks for the PR. Can you please address the failing tests before we have someone start reviewing this?

**Comment by lp4775** [comment]
Tests are failing as the personal access credential has not been implemented for the Generic Webhooks.

 This is used for authenticating post updates to a build status endpoint eg running/complete/failed. As it is generic, I will probably authenticate using the `'Authorization', 'token ` method.

I will add this functionality when I get chance, should just be a case of adding another credential type called `Generic PAT` or whatever.

**Comment by lp4775** [comment]
@shanemcd got some time to have a look at this now. Had a look at most of the tests, are there any that are known to still be failing?

**Comment by shanemcd** [comment]
Hi @lp4775 - In the `awx-api` failures above I see:

```
CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0139_generic_webhook, 0144_event_partitions in main).
```

Your migrations need to be renumbered.

In the UI tests I see:

```
2021-06-07 10:55:13.444781 | static | /awx_devel/awx/ui_next/src/screens/Template/JobTemplateDetail/JobTemplateDetail.jsx
2021-06-07 10:55:13.444854 | static |   147:12  error  'i18n' is not defined  no-undef
2021-06-07 10:55:13.444928 | static |   151:14  error  'i18n' is not defined  no-undef
2021-06-07 10:55:13.444987 | static |   153:12  error  'i18n' is not defined  no-undef
2021-06-07 10:55:13.444998 | static |
2021-06-07 10:55:13.445065 | static | /awx_devel/awx/ui_next/src/screens/Template/shared/WebhookSubForm.jsx
2021-06-07 10:55:13.445115 | static |   120:14  error  'i18n' is not defined  no-undef
```

**Comment by jbradberry** [comment]
I see two major (and probably insurmountable) problems with this:

- We are not going to support Bitbucket, since webhook requests from the free site are unsigned and it seems like poor UX to allow some bitbucket requests to break but others not.
- I am unaware of any site that allows the user to configure additional payload to add to webhook requests.  How is this part even supposed to work?

**Comment by lp4775** [comment]
@jbradberry i envisioned it working in a similar way to that of [Jenkins](https://plugins.jenkins.io/generic-webhook-trigger).

As for the config aspect, this is optional but would allow the dev to specify which variable should be used in a callback depending on the event key. This is the function that handles the override. [diff](https://github.com/ansible/awx/pull/10039/commits/3465ccc08160f23ecb609cf51ec18bdbf5352069#diff-c80eb18efdbf596480ee605607043d318310e113ec4d6215ab78c3f46ea1eaf4R243)

**Comment by lp4775** [comment]
> Hi @lp4775 - In the `awx-api` failures above I see:
> ...
> 
> Your migrations need to be renumbered.
> 


I think the migration numbering would have to be updated during a merge - otherwise we will be constantly chasing the latest migration number.

**Comment by rebeccahhh** [comment]
@lp4775 we don't change our migration numbers during merges, we do it before. This is purposeful and helps us maintain working knowledge of what types of changes will need migrations and when the appropriate time to merge them is, it also prevents any possible race condition-like merges. In other words you still have to update the migration ^_^
I know it seems tedious and perhaps one day that will change but adding that kind of functionality to our process is not a priority at this time.

File: awx/ui_next/src/screens/Template/JobTemplateDetail/JobTemplateDetail.jsx [hunk 1]
```diff
@@ -142,6 +142,17 @@ function JobTemplateDetail({ template }) {
     </TextList>
   );
 
+  function webhook_translation(alias) {
+   if (alias === 'github'){
+    return i18n._(t`GitHub`)
+    }
```
Comment (BobVanB) [inline awx/ui_next/src/screens/Template/JobTemplateDetail/JobTemplateDetail.jsx:148]:
```suggestion
    if (alias === 'github'){
      return i18n._(t`GitHub`)
    }
```

**Comment by shanemcd** [comment]
Closing due to lack of activity and number of conflicts. Please open a new PR if you would still like this to be considered.

**Pull request closed without merging**